compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -136,6 +144,242 @@ emitLinkedTuningSpec(ModuleOp module, ArrayRef<NamedSequenceOp> specsToLink) {
+  SmallVector<transform::ForeachMatchOp> foreachMatchOps;
+  // foreachMatchMap: NamedSequenceOp -> ForeachMatchOps that reference it
+  // (either as a matcher or an action). It ensures
+  // that when a NamedSequenceOp is renamed for uniqueness, the corresponding
+  // ForeachMatchOp is also updated.
+  llvm::DenseMap<transform::NamedSequenceOp, transform::ForeachMatchOp>
+      foreachMatchMap;
+
+  // Step 1: Collect NamedSequenceOps and ForeachMatchOps from inner modules.
+  for (auto innerModule : module.getBody()->getOps<ModuleOp>()) {
+    for (auto namedSequenceOp :
+         innerModule.getBody()->getOps<transform::NamedSequenceOp>()) {
+      if (namedSequenceOp.getSymName() == kKernelConfigSpecName) {
+        transform::ForeachMatchOp foreachMatch = nullptr;
+        int matchCount = 0;
+        // Iterate directly over ForeachMatchOp within kernelConfig.
+        for (auto op : namedSequenceOp.getOps<transform::ForeachMatchOp>()) {
+          if (!foreachMatch) {
+            foreachMatch = op;
+          }
+          matchCount++;
+        }
+
+        // Return failure if multiple occurrences exist.
+        if (matchCount > 1) {
+          return failure();
+        }
+        // Return failure if not foreach match op found.
+        if (!foreachMatch)
+          return failure();
+
+        foreachMatchOps.push_back(foreachMatch);
+
+        for (auto matcher : foreachMatch.getMatchers()) {
+          if (auto matcherSymRef = dyn_cast<SymbolRefAttr>(matcher)) {
+            if (auto matcherOp = dyn_cast_or_null<transform::NamedSequenceOp>(
+                    SymbolTable::lookupNearestSymbolFrom(innerModule,
+                                                         matcherSymRef))) {
+              if (!foreachMatchMap.count(matcherOp)) {
+                foreachMatchMap[matcherOp] = foreachMatch;
+              }
+            }
+          }
+        }
+        for (auto action : foreachMatch.getActions()) {
+          if (auto actionSymRef = dyn_cast<SymbolRefAttr>(action)) {
+            if (auto actionOp = dyn_cast_or_null<transform::NamedSequenceOp>(
+                    SymbolTable::lookupNearestSymbolFrom(innerModule,
+                                                         actionSymRef))) {
+              if (!foreachMatchMap.count(actionOp)) {
+                foreachMatchMap[actionOp] = foreachMatch;
+              }
+            }
+          }
+        }
+      } else {
+        namedSequenceOpsToMove.push_back(namedSequenceOp);
+      }
+    }
+  }
+
+  // Step 2-a: Ensure all ForeachMatchOps have the same result types before
+  // merging.
+  SmallVector<Type, 4> expectedResultTypes =
+      llvm::to_vector<4>(foreachMatchOps.front()->getResultTypes());
+
+  for (auto foreachMatchOp : foreachMatchOps) {
+    SmallVector<Type, 4> currentResultTypes =
+        llvm::to_vector<4>(foreachMatchOp.getResultTypes());
+
+    if (!llvm::equal(currentResultTypes, expectedResultTypes)) {
+      return failure();
+    }
```

Comment:
Instead, should we check that the result type is exactly what we expect? I think it must take any_op and return any_op.